pkg/client/k8s/v1alpha1/resource.go

```diff
@@ -0,0 +1,173 @@
+	return func(obj *unstructured.Unstructured, subresources ...string) (*unstructured.Unstructured, error) {
+		if obj == nil {
+			return nil, fmt.Errorf("nil resource instance: failed to create '%s' at namespace '%s'", gvr, namespace)
+		}
+
+		dynamic, err := NewDynamicGetter()()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create '%s' '%s' at namespace '%s'", gvr, obj.GetName(), namespace)
+		}
+
+		unstruct, err := dynamic.Resource(gvr).Namespace(namespace).Create(obj, subresources...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create '%s' '%s' at namespace '%s'", gvr, obj.GetName(), namespace)
+		}
+
+		return unstruct, nil
+	}
+}
+
+// ResourceGetter abstracts fetching an unstructured instance from kubernetes
+// cluster
+type ResourceGetter func(name string, options metav1.GetOptions, subresources ...string) (*unstructured.Unstructured, error)
+
+// NewResourceGetter returns a new instance of ResourceGetter that is capable
+// of fetching an unstructured instance from kubernetes cluster
+func NewResourceGetter(gvr schema.GroupVersionResource, namespace string) ResourceGetter {
+	return func(name string, options metav1.GetOptions, subresources ...string) (*unstructured.Unstructured, error) {
+		if len(strings.TrimSpace(name)) == 0 {
+			return nil, fmt.Errorf("missing resource name: failed to get '%s' from namespace '%s'", gvr, namespace)
+		}
+
+		dynamic, err := NewDynamicGetter()()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get '%s' '%s' from namespace '%s'", gvr, name, namespace)
+		}
+
+		unstruct, err := dynamic.Resource(gvr).Namespace(namespace).Get(name, options, subresources...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get '%s' '%s' from namespace '%s'", gvr, name, namespace)
+		}
+
+		return unstruct, nil
+	}
+}
+
+// ResourceUpdater abstracts updating an unstructured instance found in
+// kubernetes cluster
+type ResourceUpdater func(oldobj, newobj *unstructured.Unstructured, subresources ...string) (*unstructured.Unstructured, error)
+
+// NewResourceUpdater returns a new instance of ResourceUpdater that is capable
+// of updating an unstructured instance found in kubernetes cluster
+func NewResourceUpdater(gvr schema.GroupVersionResource, namespace string) ResourceUpdater {
+	return func(oldobj, newobj *unstructured.Unstructured, subresources ...string) (*unstructured.Unstructured, error) {
+		if oldobj == nil {
+			return nil, fmt.Errorf("nil old resource instance: failed to update '%s' at namespace '%s'", gvr, namespace)
+		}
+
+		if newobj == nil {
+			return nil, fmt.Errorf("nil new resource instance: failed to update '%s' at namespace '%s'", gvr, namespace)
+		}
+
+		dynamic, err := NewDynamicGetter()()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to update '%s' '%s' at namespace '%s'", gvr, oldobj.GetName(), namespace)
+		}
+
+		resourceVersion := oldobj.GetResourceVersion()
+		newobj.SetResourceVersion(resourceVersion)
```

Comment:
What are we trying to do here ? Trying to understand.

Reply:
The unstructured instance needs to have a resourceversion in-order to update.It tries to get the current resource version and sets it against the current yaml/content to proceed with update. cc @sonasingh46

Reply:
Ok.Thankyou